src/textual/widgets/_progress_bar.py

```diff
@@ -0,0 +1,432 @@
+        if percentage is None:
+            self.renderable = "--:--:--"
+        else:
+            # Keep track of the starting time.
+            if self._start_time is None:
+                self._start_time = monotonic()
+            # If we are not notified of progress for some time, periodically refresh
+            # the countdown so that the ETA is updated either way.
+            if self._refresh_timer is None:
+                self._refresh_timer = self.set_interval(1 / 2, self.update_eta)
+
+            self._refresh_timer.reset()
+            self.update_eta()
+
+    def update_eta(self) -> None:
+        """Update the ETA display."""
+        percentage = self._completion_percentage
+        # We display --:--:-- if we haven't started, if we are done,
+        # or if we don't know when we started keeping track of time.
+        if not percentage or percentage >= 1 or self._start_time is None:
+            self.renderable = "--:--:--"
+            # If we are done, we can delete the timer that periodically refreshes
+            # the countdown display.
+            if (
+                percentage is not None
+                and percentage >= 1
+                and self._refresh_timer is not None
+            ):
+                self._refresh_timer.stop()
+                self._refresh_timer = None
+        # Render a countdown timer with hh:mm:ss, unless it's a LONG time.
+        else:
+            delta = monotonic() - self._start_time
+            left = ceil((delta / percentage) * (1 - percentage))
+            minutes, seconds = divmod(left, 60)
+            hours, minutes = divmod(minutes, 60)
+            if hours > 999999:
+                self.renderable = "+999999h"
+            elif hours > 99:
+                self.renderable = f"{hours}h"
+            else:
+                self.renderable = f"{hours:02}:{minutes:02}:{seconds:02}"
+        self.refresh()
+
+
+class ProgressBar(Widget, can_focus=False):
+    """A progress bar widget."""
+
+    DEFAULT_CSS = """
+    Horizontal {
+        width: auto;
+        height: auto;
+    }
+    ProgressBar {
+        width: auto;
+        height: 1;
+    }
+    """
+
+    progress: reactive[float] = reactive(0.0)
+    """The progress so far, in number of steps."""
+    total: reactive[float | None] = reactive[Optional[float]](None)
+    """The total number of steps associated with this progress bar, when known.
+
+    The value `None` will render an indeterminate progress bar.
+    Once `total` is set to a numerical value, it cannot be set back to `None`.
+    """
+    _percentage: reactive[float | None] = reactive[Optional[float]](None)
+    """The percentage of progress that has been completed.
+
+    We keep this private to prevent the user from having multiple reactive attributes
+    that can be used to advance the progress of the bar.
+    """
+    _started = False
+    """Whether the total was set to a non-`None` value before."""
+
+    class Completed(Message):
```

Comment:
Struggling to think why we would need a message for this (ditto `Started`).

The dev tells this widget about progress, which means the dev already knows if it is complete. Don't see any value in informing them of something the probably know. Suggest we remove these messages unless you have a compelling reason to keep them...

As an aside, messages usually use present tense. It's `on_click`, not `on_clicked`. This this would be `Complete` rather than `Completed`.